File: backend/app/models/session.py
import uuid
from datetime import datetime, timezone
from sqlalchemy import (
    Column,
    String,
    Float,
    Integer,
    Boolean,
    BigInteger,
    DateTime,
    ForeignKey,
    JSON
)
from sqlalchemy.orm import relationship
from backend.app.core.database import Base

def generate_uuid():
    return str(uuid.uuid4())

class SessionModel(Base):
    __tablename__ = "sessions"

    id = Column(String(64), primary_key=True, default=generate_uuid, index=True)
    device_id = Column(String(64), index=True, default="android-device")
    exercise_name = Column(String(32), index=True, default="squat")
    started_at = Column(BigInteger, nullable=False)
    completed_at = Column(BigInteger, nullable=False)
    duration_seconds = Column(Float, nullable=False)
    total_reps = Column(Integer, nullable=False, default=0)
    valid_reps = Column(Integer, nullable=False, default=0)
    avg_peak_knee_angle = Column(Float, default=0.0)
    min_knee_angle = Column(Float, default=0.0)
    max_knee_angle = Column(Float, default=0.0)
    evidence_failure_count = Column(Integer, default=0)
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc), index=True)

    reps = relationship(
        "RepModel",
        back_populates="session",
        cascade="all, delete-orphan",
        order_by="RepModel.rep_number"
    )

    evidence_events = relationship(
        "EvidenceEventModel",
        back_populates="session",
        cascade="all, delete-orphan",
        order_by="EvidenceEventModel.start_time_ms"
    )

class RepModel(Base):
    __tablename__ = "session_reps"

    id = Column(String(64), primary_key=True, default=generate_uuid, index=True)
    session_id = Column(String(64), ForeignKey("sessions.id", ondelete="CASCADE"), nullable=False, index=True)
    rep_number = Column(Integer, nullable=False)
    is_valid = Column(Boolean, nullable=False)
    start_timestamp_ms = Column(BigInteger, nullable=False)
    peak_timestamp_ms = Column(BigInteger, nullable=False)
    end_timestamp_ms = Column(BigInteger, nullable=False)
    duration_ms = Column(BigInteger, nullable=False)
    peak_knee_angle = Column(Float, nullable=False)
    start_knee_angle = Column(Float, nullable=False)
    end_knee_angle = Column(Float, nullable=False)
    feedback_message = Column(String(255), default="")
    failure_reasons = Column(JSON, default=list)

    session = relationship("SessionModel", back_populates="reps")

class EvidenceEventModel(Base):
    __tablename__ = "session_evidence_events"

    id = Column(String(64), primary_key=True, default=generate_uuid, index=True)
    session_id = Column(String(64), ForeignKey("sessions.id", ondelete="CASCADE"), nullable=False, index=True)
    event_type = Column(String(64), nullable=False)
    start_time_ms = Column(BigInteger, nullable=False)
    end_time_ms = Column(BigInteger, nullable=False)
    duration_ms = Column(BigInteger, nullable=False)
    frame_count = Column(Integer, nullable=False, default=1)
    details = Column(String(255), default="")

    session = relationship("SessionModel", back_populates="evidence_events")

File: web/src/types/session.ts
export interface Rep {
  id: string;
  session_id: string;
  rep_number: number;
  is_valid: boolean;
  start_timestamp_ms: number;
  peak_timestamp_ms: number;
  end_timestamp_ms: number;
  duration_ms: number;
  peak_knee_angle: number;
  start_knee_angle: number;
  end_knee_angle: number;
  feedback_message: string;
  failure_reasons: string[];
}

export interface SessionSummary {
  id: string;
  device_id: string;
  exercise_name: string;
  started_at: number;
  completed_at: number;
  duration_seconds: number;
  total_reps: number;
  valid_reps: number;
  avg_peak_knee_angle: number;
  min_knee_angle: number;
  max_knee_angle: number;
  evidence_failure_count: number;
  created_at: string;
}

export interface EvidenceEvent {
  id: string;
  session_id: string;
  event_type: string;
  start_time_ms: number;
  end_time_ms: number;
  duration_ms: number;
  frame_count: number;
  details: string;
}

export interface SessionDetail extends SessionSummary {
  reps: Rep[];
  evidence_events?: EvidenceEvent[];
}

export interface SessionNlSummary {
  source: string;
  summary: string;
  disclaimer: string;
}

export interface HealthStatus {
  status: string;
  database: string;
  service: string;
  version: string;
}

export interface PlanExercise {
  id?: string;
  exercise_type: string;
  target_sets: number;
  target_reps: number;
  order_index: number;
}

export interface Plan {
  id: string;
  name: string;
  description?: string;
  is_active: boolean;
  frequency_per_week: number;
  created_at: string;
  updated_at?: string;
  exercises: PlanExercise[];
}

export interface SessionFeedback {
  id: string;
  session_id: string;
  author: string;
  message: string;
  created_at: string;
}

export interface WeeklyReport {
  report_type: string;
  start_date: string;
  completed_sessions: number;
  target_sessions: number;
  goal_met: boolean;
  total_valid_reps: number;
  total_attempted_reps: number;
  adherence_ratio_percent: number;
  total_duration_seconds: number;
  exercise_distribution: Record<string, number>;
}

export interface MonthlyReport {
  report_type: string;
  month: string;
  completed_sessions: number;
  total_valid_reps: number;
  total_attempted_reps: number;
  adherence_ratio_percent: number;
  total_therapy_hours: number;
  exercise_distribution: Record<string, number>;
}

export type ClinicianView =
  | 'OVERVIEW'
  | 'SESSIONS'
  | 'DETAIL'
  | 'REPS'
  | 'ROM'
  | 'EVIDENCE'
  | 'HISTORY'
  | 'PLANS'
  | 'REPORTS';
